### ISSUE TITLE
tar - extract dereferencing symbol links when setting permissions

### ISSUE THREAD

**Issue by Jamie-Landeg-Jones** [termux/termux-packages#1100]
As shown below, on extraction, tar attempts to set the permissions on the target of a symlink to 777.

I've tried bsd tar, gnu tar on linux, and busybox tar on android. Creation tar is irrelevent. Extraction tar only appears to do this via termux gnu tar. Please see attached example.

Cheers, Jamie

```
11:24 (9) "com.android.vending" jamie@lapcat% cat tar-error
/etc/mkshrc
.mkshrc (root)
root@vidz:/tmp #
root@vidz:/tmp # mkdir test
root@vidz:/tmp # cd test
root@vidz:/tmp/test # ls -la
total 0
drwxr-x--- 2 root root  40 Jun 21 03:11 .
drwxrwxrwt 4 root root 140 Jun 21 03:11 ..
root@vidz:/tmp/test #
root@vidz:/tmp/test # uname -a
Linux vidz.dyslexicfish.net 3.10.0 #89 SMP PREEMPT Thu Dec 10 09:51:01 CST 2015 aarch64 Android
root@vidz:/tmp/test #
root@vidz:/tmp/test # mkdir test2
root@vidz:/tmp/test # ln -s /tmp/nofile test2/soft-link
root@vidz:/tmp/test # l test2
total 0
0 drwxr-x--- 2 root root 60 Jun 21 03:13 .
0 drwxr-x--- 3 root root 60 Jun 21 03:12 ..
0 lrwxrwxrwx 1 root root 11 Jun 21 03:13 soft-link -> /tmp/nofile
root@vidz:/tmp/test # l
total 0
0 drwxr-x--- 3 root root  60 Jun 21 03:12 .
0 drwxrwxrwt 4 root root 140 Jun 21 03:11 ..
0 drwxr-x--- 2 root root  60 Jun 21 03:13 test2
root@vidz:/tmp/test #
127|root@vidz:/tmp/test # tar cvf test.tar test2
test2/
test2/soft-link
root@vidz:/tmp/test #
root@vidz:/tmp/test # l
total 12
 0 drwxr-x--- 3 root root    80 Jun 21 03:13 .
 0 drwxrwxrwt 4 root root   140 Jun 21 03:11 ..
12 -rw-r----- 1 root root 10240 Jun 21 03:13 test.tar
 0 drwxr-x--- 2 root root    60 Jun 21 03:13 test2
root@vidz:/tmp/test # rm -r test2
root@vidz:/tmp/test # ls -l
total 12
-rw-r----- 1 root root 10240 Jun 21 03:13 test.tar
root@vidz:/tmp/test #
root@vidz:/tmp/test # l
total 12
 0 drwxr-x--- 2 root root    60 Jun 21 03:14 .
 0 drwxrwxrwt 4 root root   140 Jun 21 03:11 ..
12 -rw-r----- 1 root root 10240 Jun 21 03:13 test.tar
root@vidz:/tmp/test #
root@vidz:/tmp/test # tar xvf test.tar
test2/
test2/soft-link
tar: test2/soft-link: Cannot change mode to rwxrwxrwx: No such file or directory
tar: Exiting with failure status due to previous errors
2|root@vidz:/tmp/test #
2|root@vidz:/tmp/test #
2|root@vidz:/tmp/test # l
total 12
 0 drwxr-x--- 3 root root    80 Jun 21 03:14 .
 0 drwxrwxrwt 4 root root   140 Jun 21 03:11 ..
12 -rw-r----- 1 root root 10240 Jun 21 03:13 test.tar
 0 drwxr-x--- 2 root root    60 Jun 21 03:13 test2
1|root@vidz:/tmp/test # rm -r test2
root@vidz:/tmp/test #
root@vidz:/tmp/test # busybox tar xvf test.tar
test2/
test2/soft-link
root@vidz:/tmp/test # l
total 12
 0 drwxr-x--- 3 root root    80 Jun 21 03:14 .
 0 drwxrwxrwt 4 root root   140 Jun 21 03:11 ..
12 -rw-r----- 1 root root 10240 Jun 21 03:13 test.tar
 0 drwxr-x--- 2 root root    60 Jun 21 03:14 test2
root@vidz:/tmp/test #
root@vidz:/tmp/test # rm -r test2
root@vidz:/tmp/test # l /tmp/nofile
ls: cannot access '/tmp/nofile': No such file or directory
2|root@vidz:/tmp/test #
2|root@vidz:/tmp/test # touch /tmp/nofile
root@vidz:/tmp/test # l /tmp/nofile
0 -rw-r----- 1 root root 0 Jun 21 03:15 /tmp/nofile
root@vidz:/tmp/test # busybox tar xvf test.tar
test2/
test2/soft-link
root@vidz:/tmp/test # l /tmp/nofile
0 -rw-r----- 1 root root 0 Jun 21 03:15 /tmp/nofile
root@vidz:/tmp/test # tar xvf test.tar
test2/
test2/soft-link
root@vidz:/tmp/test # l /tmp/nofile
0 -rwxrwxrwx 1 root root 0 Jun 21 03:15 /tmp/nofile
root@vidz:/tmp/test #
root@vidz:/tmp/test #
root@vidz:/tmp/test # tar --version
tar (GNU tar) 1.29
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
root@vidz:/tmp/test # ^D
```

**Comment by xeffyr** [comment]
I couldn't repeat this error:
```
[termux]:~:$ uname -a
Linux localhost 3.18.48 #1 SMP Thu Apr 13 14:32:53 UTC 2017 aarch64 Android
[termux]:~:$ tar --version
tar (GNU tar) 1.29
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
[termux]:~:$ mkdir test
[termux]:~:$ cd test
[termux]:~/test:$ ls -la
total 8
drwx------  2 u0_a150 u0_a150 4096 Jun 28 12:30 .
drwx------ 19 u0_a150 u0_a150 4096 Jun 28 12:30 ..
[termux]:~/test:$ mkdir test2
[termux]:~/test:$ ln -s /tmp/nofile test2/soft-link
[termux]:~/test:$ ls -la test2
total 8
drwx------ 2 u0_a150 u0_a150 4096 Jun 28 12:31 .
drwx------ 3 u0_a150 u0_a150 4096 Jun 28 12:30 ..
lrwxrwxrwx 1 u0_a150 u0_a150   11 Jun 28 12:31 soft-link -> /tmp/nofile
[termux]:~/test:$ tar cvf test.tar test2
test2/
test2/soft-link
[termux]:~/test:$ ls -la
total 24
drwx------  3 u0_a150 u0_a150  4096 Jun 28 12:32 .
drwx------ 19 u0_a150 u0_a150  4096 Jun 28 12:30 ..
-rw-------  1 u0_a150 u0_a150 10240 Jun 28 12:32 test.tar
drwx------  2 u0_a150 u0_a150  4096 Jun 28 12:31 test2
[termux]:~/test:$ rm -rf test2
[termux]:~/test:$ tar xvf test.tar
test2/
test2/soft-link
[termux]:~/test:$ 
```
Is I did something wrong ?

**Comment by fornwall** [comment]
@Jamie-Landeg-Jones Hm, it seems that you are running as root. In the root session, can you run `tar --version` and `which tar` and paste the output here?

**Comment by bluelightning32** [comment]
I have two different phones with termux installed. I was able to reproduce this on one, but mostly not the other. My guess is that this only reproduces on old versions of Android.

It did reproduce on:
```
$ termux-info 
Termux version:
<=0.48
Installed Termux apps:
package:com.termux  installer=null
Updatable packages:
busybox/stable 1.27.2-1 arm [upgradable from: 1.27.2]
ncurses/stable 6.0.20170923 arm [upgradable from: 6.0.20170909-1]
termux-tools/stable 0.47 all [upgradable from: 0.46]
vim/stable 8.0.1203 arm [upgradable from: 8.0.1180]
vim-runtime/stable 8.0.1203 all [upgradable from: 8.0.1180]
System information:
Linux localhost 3.10.65 #1 SMP PREEMPT Tue May 10 17:33:09 CST 2016 armv7l Android
Termux-packages arch:
arm
Android version:
5.1
Device manufacturer:
Unnecto
Device model:
U4051
```

It did not reproduce on:
```
$ termux-info
android.os.DeadObjectException: Transaction failed on small parcel; remote process probably died
	at android.os.BinderProxy.transactNative(Native Method)
	at android.os.BinderProxy.transact(Binder.java:748)
	at android.os.BinderProxy.shellCommand(Binder.java:802)
	at com.android.commands.pm.Pm.runShellCommand(Pm.java:334)
	at com.android.commands.pm.Pm.runList(Pm.java:722)
	at com.android.commands.pm.Pm.run(Pm.java:138)
	at com.android.commands.pm.Pm.main(Pm.java:107)
	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:284)
Termux version:
<=0.48
Installed Termux apps:
Updatable packages:
vim-python/stable 8.0.1203 aarch64 [upgradable from: 8.0.1180]
vim-runtime/stable 8.0.1203 all [upgradable from: 8.0.1180]
System information:
Linux localhost 3.10.73-ga51b1600b7f8 #1 SMP PREEMPT Thu Aug 17 22:07:37 UTC 2017 aarch64 Android
Termux-packages arch:
aarch64
Android version:
8.0.0
Device manufacturer:
LGE
Device model:
Nexus 5X
```

The one thing on the Android 8.0 device is that even though tar doesn't report an error, the symlink does end up with strange permissions:
```
-bash-4.4$ (umask 0; ln -s noexist sym)
-bash-4.4$ tar -cvf sym.tar sym
sym
-bash-4.4$ rm sym
-bash-4.4$ tar -xvf sym.tar
sym
-bash-4.4$ ls -ld sym
lrwx------    1 u0_a115  u0_a115          7 Oct 19 21:59 sym -> noexist
-bash-4.4$ tar -tvf sym.tar
lrwxrwxrwx u0_a115/u0_a115   0 2017-10-19 21:59 sym -> noexist
```

It appears that Android is applying the umask to the symlinks. So as a workaround, you can set the umask to 0 before running tar.

I collected strace output from the two phones. Here is the relevant part on the phone where it fails:
```
$ strace tar -xvf sym.tar
...
symlinkat("noexist", AT_FDCWD, "sym")   = 0
fstatat64(AT_FDCWD, "sym", {st_mode=S_IFLNK|0777, st_size=7, ...}, AT_SYMLINK_NOFOLLOW) = 0
utimensat(AT_FDCWD, "sym", [{tv_sec=1508475766, tv_nsec=480010362} /* 2017-10-19T22:02:46.480010362-0700 */, {tv_sec=1508475114, tv_nsec=0} /* 2017-10-19T21:51:54-0700 */], AT_SYMLINK_NOFOLLOW) = 0
fstatat64(AT_FDCWD, "sym", {st_mode=S_IFLNK|0777, st_size=7, ...}, AT_SYMLINK_NOFOLLOW) = 0
fchmodat(AT_FDCWD, "sym", 0700)         = -1 ENOENT (No such file or directory)
```

And here it is from the phone that works:
```
$ strace tar -xvf sym.tar
...
symlinkat("noexist", AT_FDCWD, "sym")   = 0
newfstatat(AT_FDCWD, "sym", {st_mode=S_IFLNK|0777, st_size=7, ...}, AT_SYMLINK_NOFOLLOW) = 0
utimensat(AT_FDCWD, "sym", [{tv_sec=1508475795, tv_nsec=647456220} /* 2017-10-19T22:03:15.647456220-0700 */, {tv_sec=1508475583, tv_nsec=0} /* 2017-10-19T21:59:43-0700 */], AT_SYMLINK_NOFOLLOW) = 0
newfstatat(AT_FDCWD, "sym", {st_mode=S_IFLNK|0777, st_size=7, ...}, AT_SYMLINK_NOFOLLOW) = 0
openat(AT_FDCWD, "sym", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 4
fchmod(4, 0700)                         = -1 EBADF (Bad file descriptor)
fcntl(4, F_GETFL)                       = 0x208000 (flags O_RDONLY|O_NOFOLLOW|O_PATH)
fchmodat(AT_FDCWD, "/proc/self/fd/4", 0700) = 0
close(4)                                = 0
```

**Comment by bluelightning32** [comment]
The difference is inside the fchmodat function. Commit [3cbc6c627fe57c9a9783c52d148078f8d52f7b96](3cbc6c627fe57c9a9783c52d148078f8d52f7b96) changed it.

Here is my test program:
```
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <errno.h>

int main(int argc, const char *argv[]) {
        int result = fchmodat(AT_FDCWD, "sym", 0700, AT_SYMLINK_NOFOLLOW);
        if (result < 0) {
                perror("fchmodat");
        } else {
                printf("fchmodat returned %d\n", result);
        }
        return 0;
}
```

Output on Android 5.1:
```
fchmodat: No such file or directory
```

Output on Android 8.0:
```
fchmodat returned 0
```

**Comment by fornwall** [comment]
@bluelightning32 Thanks a lot!

So this should be fixed running Android 6.0 or later. In general it's recommended to update to 6.0 or later if possible, as Termux works a lot better on 6.0+ and the user base on older versions is dropping - a future version of Termux will require 6.0 or later.

So I'm closing this issue, thanks for reporting @Jamie-Landeg-Jones and thanks for the help @xeffyr and @bluelightning32 !

**Comment by bluelightning32** [comment]
In case anyone else hits this error, you can work around it by scripting tar to set the umask to 0 first:
```
$ mv /data/data/com.termux/files/usr/bin/tar /data/data/com.termux/files/usr/bin/tar.real
$ cat >/data/data/com.termux/files/usr/bin/tar <<EOD
#!/data/data/com.termux/files/usr/bin/sh
umask 0
exec "${0}.real" "$@"
EOD
$ chmod u+x /data/data/com.termux/files/usr/bin/tar
```
